keps/sig-network/20200204-NetworkPolicy-verification-rearchitecture.md

```diff
@@ -0,0 +1,668 @@
+- *Dynamic scale*: In addition to increasing the performance of these tests, we also should expand their ability to evaluate CNI's with highly dynamic, realistic workloads, outputting summary metrics.
+- *Documentation and Community*: The overall situation for these tests is that they are underdocumented and poorly understood by the community, and it's not clear how these tests are vetted when they are modified; this makes it difficult for CNI providers to compare and contrast compatibility and conformance to K8s standards for NetworkPolicys.
+- *Continuous Integration*: As part of this overall effort, once this test suite is more reliable and proven to be faster, running a basic verification of it in CI with some collection of CNI providers which could feed back into upstream K8s test results would be ideal, so that we know the NetworkPolicy test and specifications, as defined, are implemented/implementable correctly by at least some CNI provider.
+
+### Goals
+
+- Rearchitect the way we write and define CNI NetworkPolicy test verifications
+- Increase the visibility and quality of documentation available for network policies
+
+#### Concrete goals
+
+For a TLDR, see https://github.com/vmware-tanzu/antrea/blob/master/hack/netpol/pkg/main/main.go of roughly what the code looks like that we are proposing to integrate.
+
+Conceptually we have 5 concrete changes that we are proposing:
+
+1. Introduce a simple "DSL" (in golang) for defining network policy's and reachability matrices.
+  - Currently, the word `DSL` might be a misnomer, but has taken hold.  Specifically, we are using a Builder API to concsiely define NetworkPolicies in a declarative way.
+  - The DSL examples are later in this document, for example: 
+    - DFL for defining test expectations:
+    ```
+	reachability := NewReachability(allPods, true).ExpectAllIngress(Pod("x/a"), false).Expect(Pod("x/a"), Pod("x/a"), true)
+    ```
+    - DSL for Defining network policies:
+    ```
+	builder := &NetworkPolicySpecBuilder{}
+	builder = builder.SetName("x", policyName).SetPodSelector(map[string]string{"pod": "a"})
+	builder.SetTypeIngress()
+
+    ```
+2. Create all namespaces and pods in the test matrix before tests start (exceptions for some tests if we want to test pod churn or label changes etc), as part of the testing library itsef.
+3. Rewrite all existing network_policy.go tests using the above DSL using (mostly the same) ginkgo descriptions as current tests do.
+4. Integrate tests with ginkgo by simply replacing existing network policy test declarations to use the new DSL
+  - putting initialization into a BeforeAll block for scaffold pods/namespaces
+5. Annotate these tests (in the code) with the CNIs they were run on, the last time they were running, and the test results (in lieu of waiting to decide on a canonical CNI)
```

Comment:
I'd say that, in general, the people changing code should be responsible for verifying it as always, with only a slightly higher bar here - since we don't have CI .  But I'm open to ideas....

Reply:
If people adding new tests are expected/required to spin up 5 or 10 clusters to run the new test against 5 or 10 plugins, that's going to discourage people from adding new tests (or result in the tests languishing un-merged in github forever).

Reply:
Yeah, I think I agree with @danwinship on this. I don't think it is reasonable to ask test contributors to test this on a variety of platforms, nor is it Kubernetes' place to call out specific external implementations. As a corollary, we don't ask that the networking tests be annotated with all the network implementations that have passed those tests.

Reply:
Ok, ill flag this as 'disagreed upon but maybe worth discussing later'. 

I think we all agree on the end goal of having provenance for breaking changes, and are maybe caught up in minutia how to get there.

Reply:
regarding dans comment: i never said 'all CNIs'.  I meant something like a comment saying

```
// Tested by abhiraut, calico 11.0, December 25th, 2021 on 4 nodes, all tests passed in 15 minutes.
``` 
which, although informal, would be extremely useful in tracking regressions (like
https://github.com/kubernetes/kubernetes/commit/55b687054d560125819e805be8eace67c9399e1f#diff-b7cdbd8f1585bcdaf1faadb3b17ac950)

Reply:
Re: @caseydavenport I see your point, but recall that  the network analogy is only 1/2 true, because many of the critical NEtworking tests run as part of Conformance in per-commit CI. (https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-kind-conformance-parallel-ipv6/1253020522829058050/) . 

The premise here is that, the problem we have, is that there *is* no analog for this b/c we dont want to make upstream tests have external CNI dependencies. 

I'd be happy to revisit ^ , and maybe suggest that a well known CNI (like calico, flannel, etc) or suite of CI's maintained by providers by included in prow, which tested netpolicies.  Ill update the KEP to that effect as well.

Reply:
(updated bullet 5 to say 'find some way to record succesfull test results, to be agreed upon later').  assuming that probably is agreeable to all, but happy to delete bullet 5 if the consensus is that no such system of record would be  agreed upon/feasible.